// @jupyterlab/application/lib/index.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
Object.defineProperty(exports, "__esModule", { value: true });
// Local CSS must be loaded prior to loading other libs.
require("../style/index.css");
const coreutils_1 = require("@jupyterlab/coreutils");
const apputils_1 = require("@jupyterlab/apputils");
const docregistry_1 = require("@jupyterlab/docregistry");
const services_1 = require("@jupyterlab/services");
const application_1 = require("@phosphor/application");
const disposable_1 = require("@phosphor/disposable");
const mimerenderers_1 = require("./mimerenderers");
const shell_1 = require("./shell");
const signaling_1 = require("@phosphor/signaling");
var layoutrestorer_1 = require("./layoutrestorer");
exports.ILayoutRestorer = layoutrestorer_1.ILayoutRestorer;
exports.LayoutRestorer = layoutrestorer_1.LayoutRestorer;
var mimerenderers_2 = require("./mimerenderers");
exports.IMimeDocumentTracker = mimerenderers_2.IMimeDocumentTracker;
var router_1 = require("./router");
exports.IRouter = router_1.IRouter;
exports.Router = router_1.Router;
var shell_2 = require("./shell");
exports.ApplicationShell = shell_2.ApplicationShell;
/**
 * JupyterLab is the main application class. It is instantiated once and shared.
 */
class JupyterLab extends application_1.Application {
    /**
     * Construct a new JupyterLab object.
     */
    constructor(options = {}) {
        super({ shell: new shell_1.ApplicationShell() });
        /**
         * A list of all errors encountered when registering plugins.
         */
        this.registerPluginErrors = [];
        this._dirtyCount = 0;
        this._busyCount = 0;
        this._busySignal = new signaling_1.Signal(this);
        this._dirtySignal = new signaling_1.Signal(this);
        this._info = Object.assign({}, JupyterLab.defaultInfo, options);
        if (this._info.devMode) {
            this.shell.addClass('jp-mod-devMode');
        }
        this.serviceManager = new services_1.ServiceManager();
        let linker = new apputils_1.CommandLinker({ commands: this.commands });
        this.commandLinker = linker;
        let registry = (this.docRegistry = new docregistry_1.DocumentRegistry());
        registry.addModelFactory(new docregistry_1.Base64ModelFactory());
        if (options.mimeExtensions) {
            let plugins = mimerenderers_1.createRendermimePlugins(options.mimeExtensions);
            plugins.forEach(plugin => {
                this.registerPlugin(plugin);
            });
        }
    }
    /**
     * Whether the application is dirty.
     */
    get isDirty() {
        return this._dirtyCount > 0;
    }
    /**
     * Whether the application is busy.
     */
    get isBusy() {
        return this._busyCount > 0;
    }
    /**
     * Returns a signal for when application changes its busy status.
     */
    get busySignal() {
        return this._busySignal;
    }
    /**
     * Returns a signal for when application changes its dirty status.
     */
    get dirtySignal() {
        return this._dirtySignal;
    }
    /**
     * The information about the application.
     */
    get info() {
        return this._info;
    }
    /**
     * Promise that resolves when state is first restored, returning layout description.
     *
     * #### Notes
     * This is just a reference to `shell.restored`.
     */
    get restored() {
        return this.shell.restored;
    }
    /**
     * Set the application state to dirty.
     *
     * @returns A disposable used to clear the dirty state for the caller.
     */
    setDirty() {
        const oldDirty = this.isDirty;
        this._dirtyCount++;
        if (this.isDirty !== oldDirty) {
            this._dirtySignal.emit(this.isDirty);
        }
        return new disposable_1.DisposableDelegate(() => {
            const oldDirty = this.isDirty;
            this._dirtyCount = Math.max(0, this._dirtyCount - 1);
            if (this.isDirty !== oldDirty) {
                this._dirtySignal.emit(this.isDirty);
            }
        });
    }
    /**
     * Set the application state to busy.
     *
     * @returns A disposable used to clear the busy state for the caller.
     */
    setBusy() {
        const oldBusy = this.isBusy;
        this._busyCount++;
        if (this.isBusy !== oldBusy) {
            this._busySignal.emit(this.isBusy);
        }
        return new disposable_1.DisposableDelegate(() => {
            const oldBusy = this.isBusy;
            this._busyCount--;
            if (this.isBusy !== oldBusy) {
                this._busySignal.emit(this.isBusy);
            }
        });
    }
    /**
     * Register plugins from a plugin module.
     *
     * @param mod - The plugin module to register.
     */
    registerPluginModule(mod) {
        let data = mod.default;
        // Handle commonjs exports.
        if (!mod.hasOwnProperty('__esModule')) {
            data = mod;
        }
        if (!Array.isArray(data)) {
            data = [data];
        }
        data.forEach(item => {
            try {
                this.registerPlugin(item);
            }
            catch (error) {
                this.registerPluginErrors.push(error);
            }
        });
    }
    /**
     * Register the plugins from multiple plugin modules.
     *
     * @param mods - The plugin modules to register.
     */
    registerPluginModules(mods) {
        mods.forEach(mod => {
            this.registerPluginModule(mod);
        });
    }
}
exports.JupyterLab = JupyterLab;
/**
 * The namespace for `JupyterLab` class statics.
 */
(function (JupyterLab) {
    /**
     * The default application info.
     */
    JupyterLab.defaultInfo = {
        name: coreutils_1.PageConfig.getOption('appName') || 'JupyterLab',
        namespace: coreutils_1.PageConfig.getOption('appNamespace'),
        version: coreutils_1.PageConfig.getOption('appVersion') || 'unknown',
        devMode: coreutils_1.PageConfig.getOption('devMode').toLowerCase() === 'true',
        deferred: { patterns: [], matches: [] },
        disabled: { patterns: [], matches: [] },
        mimeExtensions: [],
        urls: {
            page: coreutils_1.PageConfig.getOption('pageUrl'),
            public: coreutils_1.PageConfig.getOption('publicUrl'),
            settings: coreutils_1.PageConfig.getOption('settingsUrl'),
            themes: coreutils_1.PageConfig.getOption('themesUrl')
        },
        directories: {
            appSettings: coreutils_1.PageConfig.getOption('appSettingsDir'),
            schemas: coreutils_1.PageConfig.getOption('schemasDir'),
            static: coreutils_1.PageConfig.getOption('staticDir'),
            templates: coreutils_1.PageConfig.getOption('templatesDir'),
            themes: coreutils_1.PageConfig.getOption('themesDir'),
            userSettings: coreutils_1.PageConfig.getOption('userSettingsDir'),
            serverRoot: coreutils_1.PageConfig.getOption('serverRoot')
        },
        filesCached: coreutils_1.PageConfig.getOption('cacheFiles').toLowerCase() === 'true'
    };
})(JupyterLab = exports.JupyterLab || (exports.JupyterLab = {}));
